# FIXED COUPON BOND - QL EXAMPLE

This is a based on example in http://gouthamanbalaraman.com/blog/quantlib-bond-modeling.html

In [2]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
from financepy.finutils import *
from financepy.products.bonds.TuringBond import *
from financepy.market.curves.TuringDiscountCurveZeros import *

###################################################################
# FINANCEPY BETA Version 0.180 - This build: 22 Sep 2020 at 19:38 #
#     This software is distributed FREE & WITHOUT ANY WARRANTY    #
# For info and disclaimer - https://github.com/domokane/FinancePy #
###################################################################



# Define the Bond

In [8]:
issueDate = TuringDate(15,1,2010)
maturityDate = TuringDate(15, 1, 2016)
coupon = 0.06
freqType = TuringFrequencyTypes.SEMI_ANNUAL
accrualType = TuringDayCountTypes.THIRTY_360_BOND
face = ONE_MILLION

In [9]:
bond = TuringBond(issueDate, maturityDate, coupon, freqType, accrualType, face)

In [10]:
print(bond)

OBJECT TYPE: TuringBond
ISSUE DATE: FRI 15 JAN 2010
MATURITY DATE: FRI 15 JAN 2016
COUPON: 0.06
FREQUENCY: TuringFrequencyTypes.SEMI_ANNUAL
ACCRUAL TYPE: TuringDayCountTypes.THIRTY_360_BOND
FACE AMOUNT: 1000000


To see the cashflows we first need to set the settlement date of the bond. 

In [11]:
settlementDate = TuringDate(15, 1, 2015)

In [12]:
bond.printFlows(settlementDate)

THU 15 JAN 2015      30000.00 
WED 15 JUL 2015      30000.00 
FRI 15 JAN 2016    1030000.00 


## Discounting Bond Flows

We wish to define a zero rate curve. For this we need the dates and values of the zero rates.

In [13]:
zeroDates = [TuringDate(15,7,2015), TuringDate(15,1,2016)]
zeroRates = [0.005, 0.007]

In [9]:
discountCurve = TuringDiscountCurveZeros(settlementDate, zeroDates, zeroRates, 
                                      TuringFrequencyTypes.ANNUAL,
                                      TuringDayCountTypes.THIRTY_360_BOND)

In [10]:
print(discountCurve)

OBJECT TYPE: TuringDiscountCurveZeros
VALUATION DATE: THU 15 JAN 2015
FREQUENCY TYPE: TuringFrequencyTypes.ANNUAL
DAY COUNT TYPE: TuringDayCountTypes.THIRTY_360_BOND
INTERP TYPE: TuringInterpTypes.FLAT_FWD_RATES
DATES: ZERO RATES
WED 15 JUL 2015:  0.0050000
FRI 15 JAN 2016:  0.0070000



In [11]:
bond.cleanPriceFromDiscountCurve(settlementDate, discountCurve)

105.27653992490681

In [12]:
bond.fullPriceFromDiscountCurve(settlementDate, discountCurve)

105.27653992490681

As we are on the issue date of the bond there is no accrued.

This agrees with QL which finds a price of 105.27654

Copyright (c) 2020 Dominic O'Kane